From a256230d42456396f9cc9e5b999cf92eec71922e Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Thu, 28 Jul 2022 09:36:45 +0000
Subject: [PATCH] xen/arm: mpam: Introduce mpam_config for each PARTID

In MPAM initialization stage, one of the step is to reset all MSC
and RIS, where a helper function `mpam_reprogram_ris_partid()` is
used to write the reset value for each PARTID. To reuse this helper
function after boot time where we need to apply MPAM config, this
commit introduces a new `struct mpam_config` for each PARTID to
track the MPAM configuration for each partid. Also, modify the
parameter list of `mpam_reprogram_ris_partid()` and change function
`mpam_has_feature()` to a macro in this commit to keep the
consistensy with previous commits.

Since the system maximum number of PARTID is available after MPAM
hardware probing, the array of `struct mpam_config` is initialized
at the MPAM enable stage after we probed all MPAM devices. This
array is stored in each MPAM component for further query.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-5049
Change-Id: I4908a4a1b53464b8e73252bd838b7cd1486d846d
---
 xen/arch/arm/mpam.c        | 74 ++++++++++++++++++++++++++++++++++----
 xen/include/asm-arm/mpam.h | 16 ++++++---
 2 files changed, 78 insertions(+), 12 deletions(-)

diff --git a/xen/arch/arm/mpam.c b/xen/arch/arm/mpam.c
index 1a02ca92a9..2bd12e8343 100644
--- a/xen/arch/arm/mpam.c
+++ b/xen/arch/arm/mpam.c
@@ -877,7 +877,8 @@ static void mpam_reset_msc_bitmap(struct mpam_msc *msc,
         mpam_write_reg(msc, reg, bm);
 }
 
-static void mpam_reprogram_ris_partid(struct mpam_msc_ris *ris, uint16_t partid)
+static void mpam_reprogram_ris_partid(struct mpam_msc_ris *ris, uint16_t partid,
+                                      struct mpam_config *cfg)
 {
     unsigned long flags;
     struct mpam_msc *msc = ris->msc;
@@ -899,16 +900,31 @@ static void mpam_reprogram_ris_partid(struct mpam_msc_ris *ris, uint16_t partid)
                        (MPAMCFG_PART_SEL_INTERNAL | partid));
 
     if ( mpam_has_feature(mpam_feat_cpor_part, rprops) )
-        mpam_reset_msc_bitmap(msc, MPAMCFG_CPBM, rprops->cpbm_wd);
+    {
+        if ( mpam_has_feature(mpam_feat_cpor_part, cfg) )
+            mpam_write_reg(msc, MPAMCFG_CPBM, cfg->cpbm);
+        else
+            mpam_reset_msc_bitmap(msc, MPAMCFG_CPBM, rprops->cpbm_wd);
+    }
 
     if ( mpam_has_feature(mpam_feat_mbw_part, rprops) )
-        mpam_reset_msc_bitmap(msc, MPAMCFG_MBW_PBM, rprops->mbw_pbm_bits);
+    {
+        if ( mpam_has_feature(mpam_feat_mbw_part, cfg) )
+            mpam_write_reg(msc, MPAMCFG_MBW_PBM, cfg->mbw_pbm);
+        else
+            mpam_reset_msc_bitmap(msc, MPAMCFG_MBW_PBM, rprops->mbw_pbm_bits);
+    }
 
     if ( mpam_has_feature(mpam_feat_mbw_min, rprops) )
         mpam_write_reg(msc, MPAMCFG_MBW_MIN, 0);
 
     if ( mpam_has_feature(mpam_feat_mbw_max, rprops) )
-        mpam_write_reg(msc, MPAMCFG_MBW_MAX, bwa_fract);
+    {
+        if ( mpam_has_feature(mpam_feat_mbw_max, cfg) )
+            mpam_write_reg(msc, MPAMCFG_MBW_MAX, cfg->mbw_max);
+        else
+            mpam_write_reg(msc, MPAMCFG_MBW_MAX, bwa_fract);
+    }
 
     if ( mpam_has_feature(mpam_feat_mbw_prop, rprops) )
         mpam_write_reg(msc, MPAMCFG_MBW_PROP, bwa_fract);
@@ -936,7 +952,8 @@ static void mpam_reprogram_ris_partid(struct mpam_msc_ris *ris, uint16_t partid)
     spin_unlock_irqrestore(&msc->hw_lock, flags);
 }
 
-static void __init mpam_reprogram_ris(struct mpam_msc_ris *ris)
+static void __init mpam_reprogram_ris(struct mpam_msc_ris *ris,
+                                      struct mpam_config *cfg)
 {
     uint16_t partid, partid_max;
 
@@ -946,19 +963,20 @@ static void __init mpam_reprogram_ris(struct mpam_msc_ris *ris)
     partid_max = ACCESS_ONCE(mpam_partid_max);
 
     for ( partid = 0; partid < partid_max; partid++ )
-        mpam_reprogram_ris_partid(ris, partid);
+        mpam_reprogram_ris_partid(ris, partid, cfg);
 }
 
 static void __init mpam_reset_ris(struct mpam_msc_ris *ris)
 {
     struct mpam_msc *msc = ris->msc;
+    struct mpam_config empty_cfg = { 0 };
 
     BUG_ON(!spin_is_locked(&msc->lock));
 
     if ( ris->in_reset_state )
         return;
 
-    mpam_reprogram_ris(ris);
+    mpam_reprogram_ris(ris, &empty_cfg);
 }
 
 static void __init mpam_reprogram_msc(struct mpam_msc *msc)
@@ -995,6 +1013,41 @@ static void __init mpam_cpu_online(unsigned int cpu)
     spin_unlock(&mpam_list_lock);
 }
 
+static int __init __allocate_component_cfg(struct mpam_component *comp)
+{
+    if ( comp->cfg )
+        return 0;
+
+    comp->cfg = xzalloc_array(struct mpam_config, ACCESS_ONCE(mpam_partid_max));
+    if ( !comp->cfg )
+    {
+        printk(XENLOG_ERR "MPAM: Out of memory when allocating config array\n");
+        return -ENOMEM;
+    }
+
+    return 0;
+}
+static int __init mpam_allocate_config(void)
+{
+    int err = 0;
+    struct mpam_class *class;
+    struct mpam_component *comp;
+
+    BUG_ON(!spin_is_locked(&mpam_list_lock));
+
+    list_for_each_entry(class, &mpam_classes, classes_list)
+    {
+        list_for_each_entry(comp, &class->components, class_list)
+        {
+            err = __allocate_component_cfg(comp);
+            if ( err )
+                return err;
+        }
+    }
+
+    return 0;
+}
+
 /*
  * If a resource doesn't match class feature/configuration, do the right thing.
  * For 'num' properties we can just take the minimum.
@@ -1096,6 +1149,7 @@ static int __init mpam_enable(void)
 {
     struct mpam_msc *msc;
     bool all_devices_probed = true;
+    int err;
 
     spin_lock(&mpam_list_lock);
     /* Have we probed all the devices? */
@@ -1117,6 +1171,12 @@ static int __init mpam_enable(void)
     }
 
     mpam_enable_merge_features();
+    err = mpam_allocate_config();
+    if ( err )
+    {
+        spin_unlock(&mpam_list_lock);
+        return err;
+    }
 
     /* With all MSCs probed, now we can set the global variable to true. */
     mpam_enabled = true;
diff --git a/xen/include/asm-arm/mpam.h b/xen/include/asm-arm/mpam.h
index 681070eeed..d1ba5d7501 100644
--- a/xen/include/asm-arm/mpam.h
+++ b/xen/include/asm-arm/mpam.h
@@ -295,6 +295,14 @@ enum mpam_device_features {
 };
 #define MPAM_ALL_FEATURES    ((1<<MPAM_FEATURE_LAST) - 1)
 
+struct mpam_config {
+    /* Which configuration values are valid. 0 is used for reset */
+    mpam_features_t features;
+    uint32_t cpbm;
+    uint32_t mbw_pbm;
+    uint16_t mbw_max;
+};
+
 struct mpam_props
 {
     mpam_features_t features;
@@ -309,11 +317,7 @@ struct mpam_props
     uint16_t num_mbwu_mon;
 };
 
-static inline bool mpam_has_feature(enum mpam_device_features feat,
-                                    struct mpam_props *props)
-{
-    return (1 << feat) & props->features;
-}
+#define mpam_has_feature(_feat, x) ((1 << (_feat)) & (x)->features)
 
 static inline void mpam_set_feature(enum mpam_device_features feat,
                                     struct mpam_props *props)
@@ -379,6 +383,8 @@ struct mpam_component
     /* mpam_msc_ris in this component */
     struct list_head ris;
     cpumask_t affinity;
+    /* Array of configuration values, indexed by partid */
+    struct mpam_config *cfg;
     /* member of mpam_class:components */
     struct list_head class_list;
     /* parent */
-- 
2.25.1

